File: 02_parallel/problems/01_manual_partition/manual_partition.f90
program manual_partition
  use omp_lib
  integer, parameter :: n = 100
  real(8) :: a(0:n-1), s
  integer :: i, tid, nt, lo, hi
  do i = 0, n - 1
     a(i) = i + 1
  end do
  ! TODO: 下のブロックを !$omp parallel private(tid, nt, lo, hi, i, s) ... !$omp end parallel で囲み, 各スレッドが自分の担当範囲を計算するようにせよ.
  tid = omp_get_thread_num()
  nt  = omp_get_num_threads()
  lo  = tid * n / nt
  hi  = (tid + 1) * n / nt
  s = 0.0d0
  do i = lo, hi - 1
     s = s + a(i)
  end do
  print "(a,i0,a,i0,a,i0,a,i0,a,f0.6)", &
       "thread ", tid, " of ", nt, ": range [", lo, ", ", hi, "), partial sum = ", s
  ! TODO: 上で始めた parallel 領域を閉じる (!$omp end parallel).
end program manual_partition

File: 02_parallel/problems/01_manual_partition/manual_partition.cpp
#include <cstdio>
#include <omp.h>

int main() {
  const int n = 100;
  double a[n];
  for (int i = 0; i < n; i++) {
    a[i] = i + 1;
  }
  // TODO: 下のブロックの直前に #pragma omp parallel を1行追加し, 各スレッドが自分の担当範囲を計算するようにせよ.
  {
    int tid = omp_get_thread_num();
    int nt  = omp_get_num_threads();
    int lo  = tid * n / nt;
    int hi  = (tid + 1) * n / nt;
    double s = 0.0;
    for (int i = lo; i < hi; i++) {
      s += a[i];
    }
    printf("thread %d of %d: range [%d, %d), partial sum = %f\n",
           tid, nt, lo, hi, s);
  }
  return 0;
}
